File: LeetCode/cpp/1282_Group_the_People_Given_the_Group_Size_They_Belong_To.cpp
class Solution {
public:
	vector<vector<int>> groupThePeople(vector<int>& groupSizes) {
		vector<vector<int>> answer;
		unordered_map<int, vector<int>> um;

		for (int i = 0; i < groupSizes.size(); i++) {
			um[groupSizes[i]].push_back(i);
		}

		for (auto m : um) {
			int cnt = 0;
			vector<int> v;

			for (int i = 0; i < m.second.size(); i++) {
				v.push_back(m.second[i]);
				cnt++;

				if (cnt == m.first) {
					answer.push_back(v);
					cnt = 0;
					v.clear();
				}
			}
		}

		return answer;
	}
};

File: LeetCode/java/1282_Group_the_People_Given_the_Group_Size_They_Belong_To.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<List<Integer>> groupThePeople(int[] groupSizes) {
        List<List<Integer>> answer = new ArrayList<>();
        Map<Integer, List<Integer>> map = new HashMap<>();

        for (int i = 0; i < groupSizes.length; i++) {
            if (!map.containsKey(groupSizes[i])) {
                map.put(groupSizes[i], new ArrayList<>());
            }

            map.get(groupSizes[i]).add(i);
        }

        for (int key : map.keySet()) {
            int cnt = 0;
            List<Integer> list = new ArrayList<>();

            for (int i = 0; i < map.get(key).size(); i++) {
                list.add(map.get(key).get(i));
                cnt++;

                if (cnt == key) {
                    answer.add(new ArrayList<>(list));
                    cnt = 0;
                    list.clear();
                }
            }
        }

        return answer;
    }
}